src/librbd/internal.cc

```diff
@@ -2872,6 +2872,55 @@ int validate_mirroring_enabled(ImageCtx *ictx) {
             << cpp_strerror(r) << dendl;
         }
       }
+    } else if (next_mirror_mode == cls::rbd::MIRROR_MODE_DISABLED) {
+
+      bool needs_rollback = false;
+      if (current_mirror_mode == cls::rbd::MIRROR_MODE_POOL) {
+        r = cls_client::mirror_mode_set(&io_ctx, cls::rbd::MIRROR_MODE_IMAGE);
+        if (r < 0) {
+          lderr(cct) << "Failed to set mirror mode to image: "
+            << cpp_strerror(r) << dendl;
+          return r;
+        }
+        needs_rollback = true;
+      }
+
+      std::vector<std::string> image_ids;
+      r = cls_client::mirror_image_list(&io_ctx, &image_ids);
+      if (r < 0) {
+        lderr(cct) << "Failed listing images: " << cpp_strerror(r) << dendl;
+        if (needs_rollback) {
+          r = cls_client::mirror_mode_set(&io_ctx, current_mirror_mode);
+          if (r < 0) {
+            lderr(cct) << "Failed to rollback mirror mode: " << cpp_strerror(r)
+              << dendl;
+          }
+        }
+        return r;
+      }
+
+      for (const auto& img_id : image_ids) {
+        ImageCtx *img_ctx = new ImageCtx("", img_id, nullptr, io_ctx, false);
+        r = img_ctx->state->open();
+        if (r < 0) {
+          lderr(cct) << "error opening image id "<< img_id << ": "
+            << cpp_strerror(r) << dendl;
+          delete img_ctx;
+          continue;
```

Comment:
I think if you have any error disabling mirroring on an image you just need to abort -- don't want to leave the pool / images in an inconsistent state.  Same comment below.